pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -171,7 +172,15 @@ public ValidDocIndexReader getValidDocIndex() {
   @Override
   public GenericRow getRecord(int docId, GenericRow reuse) {
     // NOTE: Use PinotSegmentRecordReader to read immutable segment
-    throw new UnsupportedOperationException();
+    try {
+      PinotSegmentRecordReader pinotSegmentRecordReader = new PinotSegmentRecordReader();
```

Comment:
Please cache the record reader instead of creating one per `getRecord` call

Reply:
gotcha, thanks @Jackie-Jiang for review. Do you think i should init record reader for all immutableSegment, or should i only initialize when immutableSegment enabledUpsert?

Reply:
Just cache a record reader within this method should be good